src/psyclone/dynamo0p3.py

```diff
@@ -3934,9 +3935,34 @@ def create_halo_exchanges(self):
                                 "all dependent nodes to be halo exchanges")
                 prev_node = prev_arg_list[0].call
                 if not isinstance(prev_node, DynHaloExchange):
-                    # previous dependence is not a halo exchange so
-                    # create a new halo exchange
-                    self._add_halo_exchange(halo_field)
+                    # Previous dependence is not a halo exchange.
+                    #
+                    # There is still one case where we do not want to
+                    # add in a halo exchange. This is when the field
+                    # is continuous but only its annexed dofs are read
+                    # and the previous write is to the level 1 halo
```

Comment:
I think it is sufficient that the previous write be to owned cells since that includes annexed dofs. The level-1 halo is beyond that.

Reply:
I believe what I am doing is correct here and the comment is misleading. I'll update it. Anyway, below is the logic. Obviously let me know if I've made a mistake.

Annexed dofs only exist for continuous fields. Continuous fields can be modified in two ways 1: incremented when the parent loop iterates over cells 2: updated when the parent loop iterates over dofs. In case 1: we always (at least) iterate to the l1 halo as this ensures that annexed dofs are correct for any subsequent read. In case 2: we can currently either a) write to owned cells, which does not include annexed dofs, or b) write to a particular halo depth (or the full halo depth) via the redundant computation transformation. So, what you say would be true if we supported iterating over dofs up to annexed dofs (and perhaps will be true at some point), but is not at the moment.

Reply:
Looking at the code again, I think it would make more sense for it to be moved to the required() method for a halo exchange so I will do that before next review.